Add option to configure creating ums when moving msg to private streams with protected history.

amanagr opened the pull request:
Follow up on https://github.com/zulip/zulip/pull/16297

amanagr commented:
@timabbott there is a feature we are still missing here. For users who are subscribed to both streams but are unable to view the moved message because they are subscribed at a later date in both streams (a very common case). I think it will be best to add another option to configure this + do some cleanup in this codepath by clubbing the params we pass to do_message_update here.

timabbott commented on static/styles/zulip.css:
```diff
@@ -1440,6 +1440,7 @@ div.focused_table {
     margin: 0 5px 5px 0 !important;
     align-self: center;
     width: 100%;
+    white-space: nowrap;
```
Can you post screenshots showing what these changes do visually?

amanagr replied:
before:
<img width="499" alt="Screenshot 2020-10-03 at 9 01 49 AM" src="https://user-images.githubusercontent.com/25124304/94982317-1fe00200-0557-11eb-80f1-0616111e1e3b.png">

after:
<img width="499" alt="Screenshot 2020-10-03 at 9 01 57 AM" src="https://user-images.githubusercontent.com/25124304/94982318-25d5e300-0557-11eb-8a6e-13073b795432.png">

timabbott commented on templates/zerver/api/changelog.md:
```diff
@@ -10,6 +10,12 @@ below features are supported.
 
+**Feature level 34**
+
+* [`POST /messages`]: Add option `moved_msgs_visible_to_all_subs` to
```
I think maybe clearer wording might be something like `add_messages_to_history`?

(We avoid using `msgs` as an abbreviation anywhere in our code or API, as a sidenote -- it's valuable to have a consistent naming policy, and that for Zulip is no abbreviations).

amanagr replied:
Yes!

timabbott commented on templates/zerver/api/changelog.md:
```diff
@@ -10,6 +10,12 @@ below features are supported.
+
+* [`POST /messages`]: Add option `moved_msgs_visible_to_all_subs` to
+  set whether to create user messages when moving messages to
+  private streams with protected history.
```
We should explain this sort of feature in terms of what it means to users -- UserMessages aren't a concept for them.

"to control whether the messages should be added to the history of subscribers to the destination stream
when moving messages to a private stream with protected history".  We should have a more detailed explanation in `zulip.yaml`.

amanagr replied:
Oh! Okay!

timabbott commented on version.py:
```diff
@@ -29,7 +29,7 @@
 # Changes should be accompanied by documentation explaining what the
 # new level means in templates/zerver/api/changelog.md.
-API_FEATURE_LEVEL = 33
+API_FEATURE_LEVEL = 34
```
Need to bump this to 35 when you rebase.

amanagr replied:
Done!

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -3929,6 +3929,19 @@ paths:
+            Used when moving messages to private stream with protected history.
+            When true, this option makes the message visible to all the
+            subscribers to the stream. This will be done regardless of whether
+            the user received the messages.
```
Maybe "... the user received the messages in the old stream"

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -3929,6 +3929,19 @@ paths:
+            subscribers to the stream. This will be done regardless of whether
+            the user received the messages.
+
+            **Changes**: New in Zulip 4.0 (feature level 9).
```
9 is not what you intend here :)

amanagr replied:
Yeah 😓

timabbott commented on zerver/tests/test_events.py:
```diff
@@ -418,6 +418,7 @@ def test_stream_send_message_events(self) -> None:
                 propagate_mode,
                 False,
                 False,
+                False,
```
We should probably do a refactor that allows us to make this request more readable (a series of Falses in a row is much worse than passing these parameters via kwargs where they are named wherever they are used)

amanagr replied:
Done!

timabbott commented:
Quite so -- it looks like I merged the preparatory commits but my browser lost the comment -- anyway, the prep commits looked good, thanks @amanagr!

amanagr committed:
move_msg_to_stream: Create um for user subbed later to streams.

We were not creating um for user who was subbed later to the
old stream and new stream after message in old stream was sent.
We check which users don't have access to the msg in new stream
and create ums for those who don't.

amanagr committed:
move_msg_to_stream: Add option to configure visibility of msgs.

Add option to set whether to create UserMessages for subs with
no access to the message in the new stream with protected history.

amanagr commented on zerver/lib/actions.py:
```diff
@@ -4656,9 +4644,15 @@ def do_update_message(user_profile: UserProfile, message: Message,
+                # access this message.
                 ums_to_create = []
                 for msg in changed_messages:
+                    user_ids_for_which_ums_exist = UserMessage.objects.select_related().filter(message__id=msg.id).values_list('user_profile__id', flat=True)
```
@timabbott would this be too expensive? We need this successfully implement this feature.

amanagr replied:
Changed this into a bulk query. Should be fine.

amanagr commented:
@timabbott ping asking for another review of this. I think this fixes an important bug in our code.
<img width="2048" alt="Screenshot 2020-11-05 at 4 07 57 PM" src="https://user-images.githubusercontent.com/25124304/98230418-1e07c500-1f81-11eb-9e38-76b6ce624275.png">
<img width="2048" alt="Screenshot 2020-11-05 at 4 08 34 PM" src="https://user-images.githubusercontent.com/25124304/98230466-2fe96800-1f81-11eb-83fc-504a27638cbf.png">

timabbott commented:
I think this hasn't been rebased since the Black migration.

amanagr commented:
Testing method:
* Create 2 private streams with protected history
* Try moving any message from one stream to the other.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5700,10 +5700,11 @@ def do_update_message(
 
         if new_stream is not None:
             assert stream_being_edited is not None
+            changed_msg_ids = [msg.id for msg in changed_messages]
```
Changed this to `changed_message_ids` as part of our general no abbreviations policy.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5702,9 +5691,24 @@ def do_update_message(
+                )
                 for msg_id in changed_msg_ids:
+                    user_ids_for_which_ums_exist = [
+                        um.user_profile.id for um in ums_for_changed_msgs if um.message_id == msg_id
```
This ends up iterating through `ums_for_changed_msgs` many times, which is going to be really bad for performance.

timabbott replied:
This fixes the issue:   

```
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index 2bce1306ef..da6bfb2eb5 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -5871,17 +5871,15 @@ def do_update_message(
                 # and create new UserMessage for these users so that they can
                 # access this message.


              ums_to_create = []
-                ums_for_changed_msgs = UserMessage.objects.select_related().filter(
-                    message__id__in=changed_message_ids
-                )
+                ums_by_message_id = defaultdict(set)
+                for um in UserMessage.objects.filter(message_id__in=changed_message_ids):
+                    ums_by_message_id[um.message_id].add(um.user_profile_id)
+
                 for msg_id in changed_message_ids:
-                    user_ids_for_which_ums_exist = [
-                        um.user_profile.id for um in ums_for_changed_msgs if um.message_id == msg_id
-                    ]
                     subs_gaining_usermessages = [
                         user_id
                         for user_id in new_stream_sub_ids
-                        if user_id not in user_ids_for_which_ums_exist
+                        if user_id not in ums_by_message_id[msg_id]
                     ]
                     for user_profile_id in subs_gaining_usermessages:
                         # The fact that the user didn't have a UserMessage originally means we can infer that the user
```

timabbott commented:
I merged a couple commits ending with 8b75b6f14fe847e105e9cb0895fa094c0da32ccb, and posted a batch of comments.  I think this should be pretty close to mergeable after the above are resolved.

amanagr commented:
Still working on this.

timabbott commented:
OK.  I'm dropping this from 4.0; I believe we document the current behavior of "Grant access on move" in the /help/ documentation; is that your understanding?

amanagr commented:
We should atleast merge the first commit, it fixes the bug that if a user subscribed to two streams after the message being moved was sent, the user will not be able to see the message in new stream as we are taking difference of subscribers.

amanagr commented:
This was fixed in 7ce6095003 (4 years later :))
